### PR TITLE
cmd/libsnap-confine-private: workaround BPF memory accounting, update apparmor profile

### PR DESCRIPTION
**Pull request by bboozzoo**
Based on top of #10803. The main change is introducing `setrlimit(RLIMIT_MEMLOCK..)` for working around the old BPF object memory accounting used by pre 5.11 kernels. The problem is only visible on Debian SId, where the system uses unified hierarchy and ships a 5.10 kernel. The relevant patches are:
- https://github.com/snapcore/snapd/commit/fdb85196384373a9af8757daa863874c1f36c0a7
- https://github.com/snapcore/snapd/commit/a2fc705051ff908bb6b29ea7e8f30f756256ef98

### CODE REVIEW

**Comment by bboozzoo** [comment]
@alexmurray you probably know more about this, I see in the logs:
```
wrz 22 15:32:19 galeon audit[755574]: AVC apparmor="DENIED" operation="capable" profile="/usr/lib/snapd/snap-confine" pid=755574 comm="snap-confine" capability=39  capname="bpf"
```
which was fixed by https://github.com/snapcore/snapd/pull/10826/commits/fdb85196384373a9af8757daa863874c1f36c0a7, but then apparmor 2.x seems to choke on this and fail, so I had to revert the patch. However, I don't see any BPF operations failing, so maybe we don't need it and just live with the denial? I would really like to avoid the rabbit hole of probing apparmor_parser for support of `capability bpf` and patching snap confine apparmor profile at runtime.

File: cmd/libsnap-confine-private/device-cgroup-support.c [hunk 1]
```diff
@@ -242,6 +244,46 @@ static void _sc_cleanup_v2_device_key(sc_cgroup_v2_device_key **keyptr) {
...
+// here, for example older systemd set the limit to 64MB while newer versions
+// set it even higher). Returns the old limit setting.
+static struct rlimit _sc_cgroup_v2_adjust_memlock_limit(void) {
+    struct rlimit old_limit = {0};
+
+    sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+    if (getrlimit(RLIMIT_MEMLOCK, &old_limit) < 0) {
+        die("cannot obtain the current memlock limit");
+    }
+    /* this should be more than enough for creating the map and loading the
```
Comment (mvo5) [inline cmd/libsnap-confine-private/device-cgroup-support.c:262]: Maybe
```
//  640K ought to be enough for anybody.
const rlim_t min_memlock_limit = 640 * 1024;
```
for lolz ?

**Review by mvo5 — APPROVED** [review]
Looks good, thank you

File: cmd/libsnap-confine-private/device-cgroup-support.c [hunk 2]
```diff
@@ -242,6 +244,46 @@ static void _sc_cleanup_v2_device_key(sc_cgroup_v2_device_key **keyptr) {
...
+     * filtering program */
+    const rlim_t min_memlock_limit = 512 * 1024;
+    if (old_limit.rlim_max >= min_memlock_limit) {
+        return old_limit;
```
Comment (mardy) [inline cmd/libsnap-confine-private/device-cgroup-support.c:266]:
```
    (void)sc_set_effective_identity(old);
```
before returning :-)

**Review by mardy — CHANGES_REQUESTED** [review]
Just a small change :-)

**Review by alexmurray — CHANGES_REQUESTED** [review]
See comments in the diff - I think this can be simplified by making use of `_sc_cgroup_v2_set_memlock_limit` within `_sc_cgroup_v2_adjust_memlock_limit` itself.

**Review by mardy — APPROVED** [review]
LGTM!

**Review by alexmurray — APPROVED** [review]
LGTM - thanks :)

**Pull request merged**